ostree-ext/src/generic_decompress.rs

```
@@ -0,0 +1,227 @@
+        self.0.read(buf)
+    }
+}
+
+impl<R: Read + Send + 'static> ReadWithGetInnerMut for TransparentDecompressor<R> {
+    fn get_inner_mut(&mut self) -> &mut (dyn Read) {
+        &mut self.0
+    }
+}
+
+// GzipDecompressor
+
+struct GzipDecompressor<R: std::io::BufRead>(flate2::bufread::GzDecoder<R>);
+
+impl<R: std::io::BufRead + Send + 'static> Read for GzipDecompressor<R> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+
+impl<R: std::io::BufRead + Send + 'static> ReadWithGetInnerMut for GzipDecompressor<R> {
+    fn get_inner_mut(&mut self) -> &mut (dyn Read) {
+        self.0.get_mut()
+    }
+}
+
+// ZstdDecompressor
+
+struct ZstdDecompressor<'a, R: std::io::BufRead>(zstd::stream::read::Decoder<'a, R>);
+
+impl<'a: 'static, R: std::io::BufRead + Send + 'static> Read for ZstdDecompressor<'a, R> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+
+impl<'a: 'static, R: std::io::BufRead + Send + 'static> ReadWithGetInnerMut
+    for ZstdDecompressor<'a, R>
+{
+    fn get_inner_mut(&mut self) -> &mut (dyn Read) {
+        self.0.get_mut()
+    }
+}
+
+pub(crate) struct Decompressor {
+    inner: Box<dyn ReadWithGetInnerMut>,
+    finished: bool,
+}
+
+impl Read for Decompressor {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.inner.read(buf)
+    }
+}
+
+impl Drop for Decompressor {
+    fn drop(&mut self) {
+        if self.finished {
+            return;
+        }
+
+        // We really should not get here; users are required to call
+        // `finish()` to clean up the stream.  But we'll give
+        // best-effort to clean things up nonetheless.  If things go
+        // wrong, then panic, because we're in a bad state and it's
+        // likely that we end up with a broken pipe error or a
+        // deadlock.
+        self._finish().expect("Decompressor::finish MUST be called")
```

(Sorry this is an unrelated pre-existing issue, but I'm noticing it as it's revealed by the reported bug; we can add another commit on top here, or just do another PR)

In error paths, we still expect `finish()` to not be called, right?

Bigger picture, this whole thing gets into [linear types](https://faultlore.com/blah/linear-rust/) if we were to really try to enforce it, and it's basically not possible today.

As is basically what we've done is removed all real possibility for errors from this invocation as it should boil down to `read()` from a pipe fd which should basically never fail.

That said still it's not clear to me that we should have ever been panic'ing on this error at all.

So bottom line IMO:

- Change the error message to `Failed to flush pipe`
- Change it to an explicit `// Ignore errors here, shouldn't fail let _ = self.finish(); ` (or maybe `tracing::warn!` on the error?)

---

Yeah in error paths we're going to short-circuit return and then the decompressor will be dropped.

I'm leaning towards still panic'ing and changing the error message to `Failed to flush pipe`.  At this point I *think* it's basically impossible to ever hit that panic... but if somebody somewhere *does* hit it I'd like it to be very loud and obvious so we have a chance to maybe hear about it (much like the report that spawned this PR).  I'm afraid if we ignore the error and log somewhere that nobody will ever see, we're going to have instances where things go back to hanging/deadlocking and then we'll have an even worse experience.

---

Yep OK by me to just change the error message.

---

That said on all of this the *real* fix is for us to switch to GetRawBlob - I'll file an issue

---

Added a commit to address this